### PR TITLE
add check each sc instance health

### PR DESCRIPTION
**Pull request by daqingshu**
for issue #594
please review, thanks

### CODE REVIEW

**Comment by tianxiaoliang** [comment]
Can one of the admins verify this patch?

File: pkg/scclient/client.go [hunk 1]
```diff
@@ -981,3 +993,32 @@ func (c *RegistryClient) startBackOff(microServiceID string, callback func(*Micr
...
+			c.Config.Status[v] = unavailable
+		} else {
+			c.Config.Status[v] = available
+			conn.Close()
+		}
+	}
+}
+
+func (c *RegistryClient) sciMonitor() {
+	ticker := time.NewTicker(5 * time.Second)
```
Comment (tianxiaoliang) [inline pkg/scclient/client.go]: 5s is so frequency

File: pkg/scclient/client.go [hunk 2]
```diff
@@ -954,12 +965,13 @@ func (c *RegistryClient) WatchMicroService(microServiceID string, callback func(
...
 func (c *RegistryClient) getAddress() string {
-	next := RoundRobin(c.Config.Addresses)
-	addr, err := next()
-	if err != nil {
-		return DefaultAddr
+	for _, v := range c.Config.Addresses {
+		if c.Config.Status[v] == unavailable {
+			continue
+		}
+		return v
```
Comment (tianxiaoliang) [inline pkg/scclient/client.go]: after you filter ths address list ,you shoud use round robin to select one

File: pkg/scclient/client.go (hunk 1 above)
Comment (daqingshu) [reply pkg/scclient/client.go]: how long is appropriate, 25s ? 30s ?

Comment (tianxiaoliang) [reply pkg/scclient/client.go]: imagine you got 1000 service instances, each one dial each sc server instance's port every 5 seconds that is 1000*sc instances connections．
default should be 30s
and can be configurable

Comment (daqingshu) [reply pkg/scclient/client.go]: agree, but have some trouble
1. if can be configurable, should pass the option with r.Initialize(sco)
2. I found there are one service instance may create three RegistryClient instances "r.Initialize(sco)"

File: pkg/scclient/client.go (hunk 2 above)
Comment (daqingshu) [reply pkg/scclient/client.go]: should I not return v, just create a temp slice store v, then use origin code pass it to RoundRobin()?

Comment (tianxiaoliang) [reply pkg/scclient/client.go]: sure

File: pkg/scclient/client.go (hunk 1 above)
Comment (tianxiaoliang) [reply pkg/scclient/client.go]: ok， use env “CHASSIS_SC_HEADLTH_CHECH_INTERVAL=30s”

File: pkg/scclient/client.go (hunk 2 above)
Comment (daqingshu) [reply pkg/scclient/client.go]: done

File: pkg/scclient/client.go [hunk 3]
```diff
@@ -40,6 +42,8 @@ const (
 	DefaultRetryTimeout = 500 * time.Millisecond
 	HeaderRevision      = "X-Resource-Revision"
 	EnvProjectID        = "CSE_PROJECT_ID"
+	//EnvCheckSCIInterval sc instance health check interval in second
+	EnvCheckSCIInterval = "CHASSIS_SC_HEALTH_CHECK_INTERVAL"
```
Comment (tianxiaoliang) [inline pkg/scclient/client.go]: https://docs.go-chassis.com/user-guides/env.html add env in this doc

File: pkg/scclient/client.go [hunk 4]
```diff
@@ -981,3 +1004,43 @@ func (c *RegistryClient) startBackOff(microServiceID string, callback func(*Micr
 		return
 	}
 }
+
+func (c *RegistryClient) checkSCIHealth() {
```
Comment (tianxiaoliang) [inline pkg/scclient/client.go]: plz add UT for checkSCIHealth

File: pkg/scclient/client.go [hunk 5]
```diff
@@ -954,7 +969,15 @@ func (c *RegistryClient) WatchMicroService(microServiceID string, callback func(
 }
 
 func (c *RegistryClient) getAddress() string {
```
Comment (tianxiaoliang) [inline pkg/scclient/client.go]: plz add UT for getAddress

File: pkg/scclient/client.go (hunk 3 above)
Comment (daqingshu) [reply pkg/scclient/client.go]: done.

File: pkg/scclient/client.go (hunk 4 above)
Comment (daqingshu) [reply pkg/scclient/client.go]: wait..

**Comment by daqingshu** [comment]
temp close.

**Pull request closed without merging**